### PR TITLE
feat : send new leave request mail

### PR DESCRIPTION
**Pull request by cp-mansi-d**
## Review checklist

**Note:** Make sure all points should be checked before submiting PR for review, otherwise PR will be cosidered invalid.

- [x] Follow guidelines for creating branch name, adding commit messages or PR title
- [x] Add MR description if applicable
- [x] Follow guidelines for naming conventions everywhere (i.e files/folders, data structures)
- [x] Reuse code (if the same code is written twice, make it common and reuse it at both places)
- [x] Remove unused or commented code if not required
- [x] Make MR self-approved - review MR as a reviewer and give it self-approval if everything is ok, if not make the required changes
- [x] In progress MR should be marked as draft
- [x] Make MR mark as ready before submitting it for review
- [x] Add images, videos, or gifs of UI 
- [ ] Add page speed score screenshot of UI (It should be > 85 for mobile and > 90 for desktops)

**Design should be tested on:**
- [ ] Inspector (all iPhones and android devices like samsung, motorola)
- [ ] Desktop (chrome, firefox, safari)
- [ ] Laptop / Macbook  (chrome, firefox, safari)
- [ ] iPhone(safari)
- [ ] Available android devices  (chrome, firefox)
- [ ] iPad(safari)
- [ ] Tablet (chrome, firefox)

**Description**
- create API for send mail for new leave request.
- Get data from Unity.
New leave request - mansi d@canopas com - Canopas Mail

### CODE REVIEW

File: api-doc.md [hunk 1]
```diff
@@ -298,3 +298,42 @@
...
+```
+- Method : POST
+- Endpoint : /api/new-leave-mail
+- Description : API for sending mail to HR for New leave request
+- Request:
+    - Headers : none
+    - Body :
+        {
+            "name": "mansi",
+            "date": "3 jan",
```
Comment (nidhi-canopas) [inline api-doc.md:313]: Should it contain year as well?

File: api-doc.md [hunk 2]
```diff
@@ -298,3 +298,42 @@
...
+
+```
+- Method : POST
+- Endpoint : /api/new-leave-mail
+- Description : API for sending mail to HR for New leave request
+- Request:
+    - Headers : none
+    - Body :
+        {
+            "name": "mansi",
```
Comment (nidhi-canopas) [inline api-doc.md:312]: Should it have surname as well, in case we have 2 members with same name?

File: main.go [hunk 3]
```diff
@@ -54,6 +56,8 @@ func setupRouter(sqlDb *sqlx.DB) *gin.Engine {
 
 	router.GET("/api/github/stars", contribution.GetStargazers)
 
+	router.POST("/api/new-leave-mail", unityRepo.SendNewLeaveMail)
```
Comment (nidhi-canopas) [inline main.go:59]: I think new is not required here, if no existing leave mail is there.

File: api-doc.md [hunk 4]
```diff
@@ -298,3 +298,42 @@
...
+```
+- Method : POST
+- Endpoint : /api/leave/new
+- Description : API for sending mail to HR for New leave request
+- Request:
+    - Headers : none
+    - Body :
+        {
+            "name": "mansi dhameliya",
+            "date": "3 jan 2023",
```
Comment (cp-sumi-k) [inline api-doc.md:313]: How to send date for multiple leaves, like leave for 3 days?

File: api-doc.md [hunk 5]
```diff
@@ -298,3 +298,42 @@
...
+            "status": "request",
```
Comment (cp-sumi-k) [inline api-doc.md:314]: Any need of status?

File: api-doc.md [hunk 6]
```diff
@@ -298,3 +298,42 @@
...
+            "receiver": "mansi@gmail.com",
```
Comment (cp-sumi-k) [inline api-doc.md:315]: Shouldn't receiver be HR?

File: api-doc.md [hunk 7]
```diff
@@ -298,3 +298,42 @@
...
+    - If any server error is there then,
+    - Status Code: 500 Internal server error
+    - Data: no response data
+
+    - If request will success ,
+    - Status Code: 200 Ok
+    - Headers : none
+    - Data :
+               {
+            "name": "mansi dhameliya",
```
Comment (cp-sumi-k) [inline api-doc.md:334]: I think, no need for this response, we can send messages like, `Leave applied successfully` or `Leave request has been sent successfully`.

File: unity/unity.go [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/gin-gonic/gin"
+	"github.com/gin-gonic/gin/binding"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	CHARSET = "utf-8"
+)
+
+type UnityLeaveData struct {
```
Comment (cp-sumi-k) [inline unity/unity.go:21]: Rename it as `LeaveData`

File: unity/unity.go [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+	Name     string `json:"name" form:"name"`
+	Date     string `json:"date" form:"date"`
```
Comment (cp-sumi-k) [inline unity/unity.go:23]: Need to take 2 dates, `start_date` and `end_date`

File: unity/unity.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+	Status   string `json:"status" form:"status"`
+	Receiver string `json:"receiver" form:"receiver"`
+}
+
+type Template struct {
```
Comment (cp-sumi-k) [inline unity/unity.go:28]: Let's rename it to `LeaveRepository`

File: unity/unity.go [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+func New(templateFs embed.FS, utilsRepo utils.UtilsRepository) *Template {
+	templates, _ := template.ParseFS(templateFs, "templates/*.html")
+
+	return &Template{
+		templates: templates, UtilsRepo: utilsRepo,
+	}
+}
+
+func (repository *Template) SendNewLeaveMail(c *gin.Context) {
```
Comment (cp-sumi-k) [inline unity/unity.go:41]: Rename function name to `SendLeaveRequest`

File: unity/unity_test.go [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+func (faker *stubUtilsRepo) SaveJobsToSpreadSheet(input []string) {
+	/** this is stub method for adding jobs details in google spreadsheet */
+}
+
+func Test_init(t *testing.T) {
+	repo, err = initializeRepo()
+	assert.Nil(t, err)
+	testRequests = []utils.TestRequest{
```
Comment (cp-sumi-k) [inline unity/unity_test.go:53]: Let's write a unit test function-wise. Please refer to `blogs_test`

**Review by cp-sumi-k — CHANGES_REQUESTED** [review]

File: unity/go.mod [hunk 13]
```diff
@@ -0,0 +1,58 @@
+module unity
```
Comment (cp-sumi-k) [inline unity/go.mod:1]: Rename the directory name to `leave` instead of `unity`

**Review by cp-sumi-k — CHANGES_REQUESTED** [review]

File: api-doc.md (hunk 4 above)
Comment (cp-mansi-d) [reply api-doc.md:313]: Sir told me that only one date will be used for the date of **application(applied date)** of display leave request, and other information will be displayed in unity.

File: api-doc.md (hunk 5 above)
Comment (cp-mansi-d) [reply api-doc.md:314]: for display status : request,approve,reject,etc

**Review by nidhi-canopas — DISMISSED** [review]

File: unity/unity.go (hunk 9 above)
Comment (cp-mansi-d) [reply unity/unity.go:23]: have you discuss with jimmy sir that should we take two dates?

**Review by cp-sumi-k — APPROVED** [review]

**Pull request merged**